File: chatbot/src/main_agent/qa.py
from dotenv import load_dotenv

load_dotenv()

from prompt import QA_GENERATOR_TEMPLATE, QA_EVALUATION_TEMPLATE, RESPONSE_JSON_QA
from agent_state import AgentState

from langchain_core.prompts import ChatPromptTemplate
from langchain_core.messages import AIMessage

from langchain_openai import ChatOpenAI

def generate_qa(state: AgentState):
    print("Entering generate_qa")
    text = state["related_subject_content"]
    number = 3
    tone = "difficult"        
    subject = state["related_subject"]
    rephrased_question = state["rephrased_question"]
    language = state["language"]
    llm = ChatOpenAI(
        model="gpt-4o-mini",
    )
    qa_generator_prompt = ChatPromptTemplate.from_template(QA_GENERATOR_TEMPLATE)
    qa_evaluation_prompt = ChatPromptTemplate.from_template(QA_EVALUATION_TEMPLATE)
    qa_chain = qa_generator_prompt | llm
    response = qa_chain.invoke(
        {"text": text, "number": number, "tone": tone, "subject": subject, "response_json": RESPONSE_JSON_QA, "question": rephrased_question, "language": language}
    )
    generated_qa = response.content.strip()
    state["messages_"].append(AIMessage(content=generated_qa))

    qa_review_chain = qa_evaluation_prompt | llm
    response = qa_review_chain.invoke(
        {"tone": tone, "subject": subject, "quiz": generated_qa, "language": language}
    )
    generated_review = response.content.strip()
    state["messages_"].append(AIMessage(content=generated_review))

    print(f"generate_qa: rephrased_question: {state["rephrased_question"]}")
    print(f"generate_qa: Generated Q&A: {generated_qa}")
    print(f"generate_qa: Generated review: {generated_review}")
    return state

File: Backend/src/chatbot/main-agent/qa.ts
import { Injectable } from '@nestjs/common';
import { QA_GENERATOR_TEMPLATE, QA_EVALUATION_TEMPLATE, RESPONSE_JSON_QA } from './prompt';
import { ChatOpenAI } from '@langchain/openai';
import { ChatPromptTemplate } from '@langchain/core/prompts';
import { AIMessage } from '@langchain/core/messages';
import { ConfigService } from '@nestjs/config';
import AgentState from './agent_state';

@Injectable()
export class QaService {
  private readonly llm: ChatOpenAI;

  constructor(private configService: ConfigService) {
    this.llm = new ChatOpenAI({
      model: "gpt-4o-mini",
      apiKey: this.configService.get<string>('OPENAI_API_KEY'),
    });
  }

  async generateQa(state: typeof AgentState.State): Promise<Partial<typeof AgentState.State>> {
    console.log("Entering generateQa");

    const text = state.related_subject_content;
    const number = 3;
    const tone = "difficult";
    const subject = state.related_subject;
    const rephrased_question = state.rephrased_question;
    const language = state.language;

    const qa_generator_prompt = ChatPromptTemplate.fromTemplate(QA_GENERATOR_TEMPLATE);
    const qa_evaluation_prompt = ChatPromptTemplate.fromTemplate(QA_EVALUATION_TEMPLATE);

    try {
      const qa_chain = qa_generator_prompt.pipe(this.llm);
      const response = await qa_chain.invoke({
        text,
        number,
        tone,
        subject,
        response_json: RESPONSE_JSON_QA,
        question: rephrased_question,
        language
      });

      const generated_qa = response.text.trim();
      state.messages_.push(new AIMessage(generated_qa));

      const qa_review_chain = qa_evaluation_prompt.pipe(this.llm);
      const reviewResponse = await qa_review_chain.invoke({
        tone,
        subject,
        quiz: generated_qa,
        language
      });

      const generated_review = reviewResponse.text.trim();
      state.messages_.push(new AIMessage(generated_review));

      console.log(`generateQa: rephrased_question: ${state.rephrased_question}`);
      console.log(`generateQa: Generated Q&A: ${generated_qa}`);
      console.log(`generateQa: Generated review: ${generated_review}`);
    } catch (error) {
      console.error(`Error generating Q&A: ${error}`);

      // Fallback response in case of error
      state.messages_.push(new AIMessage(
        "I couldn't generate the Q&A questions at the moment. Please try again later."
      ));
    }

    return state;
  }
}